sdk/core/System.Net.ClientModel/src/Options/PipelineOptions.cs

```diff
@@ -5,34 +5,142 @@ namespace System.Net.ClientModel.Core;
+    private PipelinePolicy? _retryPolicy;
+    private PipelineTransport? _transport;
 
-    #endregion
+    private string? _serviceVersion;
+    private TimeSpan? _networkTimeout;
+    private MessageClassifier? _messageClassifier;
 
-    #region Pipeline creation: Required policy overrides
+    public PipelineOptions()
+    {
+    }
 
-    public PipelinePolicy? RetryPolicy { get; set; }
+    public virtual MessagePipeline GetPipeline()
+    {
+        if (_pipeline != null) return _pipeline;
 
-    public PipelinePolicy? LoggingPolicy { get; set; }
+        lock (_lock)
+        {
+            if (IsFrozen)
+            {
+                if (_pipeline == null)
+                {
+                    throw new InvalidOperationException("Unexpected state: null pipeline and frozen options.");
+                }
 
-    public PipelineTransport? Transport { get; set; }
+                return _pipeline;
+            }
 
-    #endregion
+            _pipeline = MessagePipeline.Create(this);
+
+            Freeze();
+
+            return _pipeline;
+        }
+    }
+
+    private bool IsFrozen => _isFrozen;
+
+    private void Freeze() => _isFrozen = true;
+
+    #region Pipeline creation options
 
-    #region Pipeline creation: Policy-specific settings
+    // Note that all properties on PipelineOptions are nullable.
+    // This gives us the ability to understand whether a caller passed them
+    // as input or whether anything we add/set on PipelineOptions was set
+    // internally.  If a property on PipelineOptions has a null value initially,
+    // we will set a default value for it when options is frozen.
 
-    // TODO: when do we want these to be nullable and when do we want to set defaults?  Why?
-    public TimeSpan? NetworkTimeout { get; set; }
+    public PipelinePolicy[]? PerCallPolicies
+    {
+        get => _perCallPolicies;
+        set
+        {
+            AssertNotFrozen();
+            _perCallPolicies = value;
+        }
+    }
 
-    public virtual MessageClassifier? MessageClassifier { get; set; }
+    public PipelinePolicy[]? PerTryPolicies
+    {
+        get => _perTryPolicies;
+        set
+        {
+            AssertNotFrozen();
+            _perTryPolicies = value;
+        }
+    }
+
+    public PipelinePolicy? RetryPolicy
+    {
+        get => _retryPolicy;
+        set
+        {
+            AssertNotFrozen();
+            _retryPolicy = value;
+        }
+    }
+
+    public PipelineTransport? Transport
+    {
+        get => _transport;
+        set
+        {
+            AssertNotFrozen();
+            _transport = value;
+        }
+    }
+
+    public string? ServiceVersion
+    {
+        get => _serviceVersion;
+        set
+        {
+            AssertNotFrozen();
+            _serviceVersion = value;
+        }
+    }
+
+    public TimeSpan? NetworkTimeout
+    {
+        get => _networkTimeout;
+        set
+        {
+            // This one doesn't freeze with the pipeline
```

Comment:
We don't want to make any options on PipelineOptions not-freezable.

1. Hard to explain to users/ppl adding options which falls in which category
2. Creates problem where you can't ever move code that depends on these settings into the pipeline or out of the pipeline

Reply:
Addressed - all properties on PipelineOptions are now freezable.